Sources/Core/UserProfileService.swift

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
+//
+// You are hereby granted a non-exclusive, worldwide, royalty-free license to use,
+// copy, modify, and distribute this software in source code or binary form for use
+// in connection with the web services and APIs provided by Facebook.
+//
+// As with any software that integrates with the Facebook platform, your use of
+// this software is subject to the Facebook Developer Principles and Policies
+// [http://developers.facebook.com/policy/]. This copyright notice shall be
+// included in all copies or substantial portions of the software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+// FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+// COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+// IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+// CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+import Foundation
+
+class UserProfileService {
+  private let oneDayInSeconds = TimeInterval(60 * 60 * 24)
+  private(set) var graphConnectionProvider: GraphConnectionProviding
+  private(set) var logger: Logging
+  private(set) var notificationCenter: NotificationObserving & NotificationPosting
+
+  private(set) var userProfile: UserProfile?
+
+  var shouldUpdateOnAccessTokenChange: Bool = false {
+    didSet {
+      switch shouldUpdateOnAccessTokenChange {
+      case true:
+        notificationCenter.addObserver(
+          self,
+          selector: #selector(refresh),
+          name: .FBSDKAccessTokenDidChangeNotification,
+          object: nil
+        )
+
+      case false:
+        notificationCenter.removeObserver(self)
+      }
+    }
+  }
+
+  private var isCurrentProfileOutdated: Bool {
+    guard let profile = userProfile else {
+      return true
+    }
+
+    return Date().timeIntervalSince(profile.fetchedDate) > oneDayInSeconds
+  }
+
+  init(
+    graphConnectionProvider: GraphConnectionProviding = GraphConnectionProvider(),
+    logger: Logging = Logger(),
+    notificationCenter: NotificationObserving & NotificationPosting = NotificationCenter.default
+    ) {
+    self.graphConnectionProvider = graphConnectionProvider
+    self.logger = logger
+    self.notificationCenter = notificationCenter
+  }
+
```

Comment:
Need to clean up

```suggestion

deinit {
  notificationCenter.removeObserver(self)
}
```

Reply:
I’ll double check we need this if we use block syntax.

Reply:
So based on this: https://developer.apple.com/library/archive/releasenotes/Foundation/RN-FoundationOlderNotes/index.html#X10_11Notes
If we use block syntax for notifications we need to clean them up, otherwise the system handles cleanup automatically. (as of iOS 9)

Reply:
Spoke in person, resolving based on discussion.